In [5]:
#Problem 6
# Solution
def factorial(num):
    if num==0:
        return 1
    return num * factorial(num-1)

num = int(input("Enter Number:"))
fac = factorial(num)
print(f"Factorial of Number: {num} = {fac}")

Enter Number: 5


Factorial of Number: 5 = 120


In [8]:
# Problem 5
# Solution
numbers = [5,2,12,1,45,33]
largest = max(numbers)
print("The maximum value is:",largest)

largest = numbers[0]
for x in numbers:
    if x > largest:
        largest = x

print("The maximum value is:",largest)


The maximum value is: 45
The maximum value is: 45


In [17]:
# Problem 7
# Solution
file = None
try:
    with open("Day2/sample.txt","r") as file:
        content = file.read()
        print(content)
except FileNotFoundError:
    print("File not found")    


Python is a multi-paradigm programming language. Object-oriented programming and structured programming are fully supported, and many of their features support functional programming and aspect-oriented programming – including metaprogramming[57] and metaobjects.[58] Many other paradigms are supported via extensions, including design by contract[59][60] and logic programming.[61] Python is often referred to as a 'glue language'[62] because it is purposely designed to be able to integrate components written in other languages.

Python uses dynamic typing and a combination of reference counting and a cycle-detecting garbage collector for memory management.[63] It uses dynamic name resolution (late binding), which binds method and variable names during program execution.


In [27]:
# Problem 8
# Solution
file = None
try:
    file = open("Day2/sample.txt","r")
    
    print("File exist")
    
except FileNotFoundError:
    print("Unable to read file")


import os
if os.path.exists("Day2/sample.txt"):
    print("File Exist")
else:
    print("File doesnot exist")

from pathlib import Path
file_path = Path("Day2/sample.txt")
if file_path.exists():
    print("File Exists")
else:
    print("File doesnot exist")

File exist
File Exist
File Exists


In [30]:
# Problem 9
# Solution
# 2x + y = 5
# x - y = 1
# x = y+1
# 2(y+1)+y = 5
# 2y+2+y = 5
# 3y = 3
# y = 1
# x = 2

import numpy as np
# Coefficients of Equations
# 2X+y = 5 -> [2,1]
# x-y = 1 ->[1,-1]

# A system of equations can be written as:
#    A . X = B
# Where 
# - A = matrix of Coefficient 
# - X = column vector of unknowns
# - B = column vector of constants
# So, for our system:
#    2   1       x      5
#    1  -1       y      1

# That means:
#  A =  2   1
#       1  -1

#  B =  5
#       1


A = np.array([[2,1],[1,-1]])
B = np.array([5,1])

x,y = np.linalg.solve(A,B)

print(f"x={x},y={y}")


x=2.0,y=1.0


In [1]:
# Program 10
def f(x):
    return x**2

analytical = (1**3)/3 - (0**3)/3

def integrate_trapezoidal(a,b,n):
    h=(b-a)/n # step size
    result = 0.5 *(f(a)+f(b))
    for i in range(1,n):
        result+=f(a+i*h)
    result*=h
    return result

a, b = 0, 1
n = 1000

numerical = integrate_trapezoidal(a,b,n)
print(f"Analytical result: {analytical:.6f}")
print(f"Numerical (Trapezoidal) result: {numerical:.6f}")

Analytical result: 0.333333
Numerical (Trapezoidal) result: 0.333333


"""
1️⃣ What Is Integration?

Integration is a mathematical operation that finds the area under a curve.

If you have a function 
𝑓(𝑥)
f(x), then the definite integral

∫𝑎
𝑏𝑓(𝑥)𝑑𝑥
∫
a
b
	​

f(x)dx

represents the total area under the curve of f(x) from x=a to x=b.
𝑓(𝑥)
f(x): the function you’re integrating

𝑎
a, 
𝑏
b: lower and upper limits of integration

𝑑
𝑥
dx: a tiny change in 
𝑥
x (width of an infinitesimal strip)

The idea is to add up infinitely many thin rectangles under the curve.

📘 2️⃣ Your Function: 

f(x)=x^2

This function is a simple parabola:

f(x)=x^2

It passes through points like:

(0, 0)

(0.5, 0.25)

(1, 1)

If you plot it between 0 and 1, you get a curved shape rising upward.

🧩 3️⃣ Integration of 
𝑓
(
𝑥
)
=
𝑥
2
f(x)=x
2

To find the area under the curve from 0 to 1:

∫
0
1
𝑥
2
 
𝑑
𝑥
∫
0
1
	​

x
2
dx
Step 1: Find the antiderivative

The antiderivative of 
𝑥
2
x
2
 is:

(𝑥^3)/3+𝐶


(where C is a constant, not needed for definite integrals)

Step 2: Apply the limits (0 and 1)



✅ So the area under the curve from 0 to 1 is 1/3 ≈ 0.3333.

💻 4️⃣ How It Works in the Program

In the code:

def f(x):
    return x ** 2


Numerically (approximation): using small rectangles or trapezoids to sum the area.

Example (trapezoidal rule):

def integrate_trapezoidal(a, b, n):
    h = (b - a) / n
    result = 0.5 * (f(a) + f(b))
    for i in range(1, n):
        result += f(a + i * h)
    result *= h
    return result


Here:

h = width of each small strip.

f(a + i*h) = height of each strip.

It adds all strip areas ≈ total area under the curve.

📊 5️⃣ Visualization (Conceptually)
|
|       *
|     *     f(x) = x²
|   * 
| * 
+----------------------
0        0.5         1


Each small trapezoid’s area adds up to approximately 1/3.



Would you like me to show a plot of the function and shaded area under the curve to visualize this integration?
"""

In [4]:
# Problem 11
import numpy as np
A = np.array([[2,1],[5,3]])
print("Original Matrix: A:")
print(A)

det_A = np.linalg.det(A)
print("\nDeterminant of A:",det_A)

if det_A !=0:
    inv_A = np.linalg.inv(A)
    print("\nInverse of A:")
    print(inv_A)
else:
    print("\n Matrix is singular - inverse doesnot exist.")

Original Matrix: A:
[[2 1]
 [5 3]]

Determinant of A: 1.0000000000000002

Inverse of A:
[[ 3. -1.]
 [-5.  2.]]


In [5]:
A = [[1,2,3],[0,4,5],[1,0,6]]
det_A = np.linalg.det(A)

In [6]:
det_A

np.float64(22.000000000000004)

In [8]:
inv_A = np.linalg.inv(A)
inv_A

array([[ 1.09090909, -0.54545455, -0.09090909],
       [ 0.22727273,  0.13636364, -0.22727273],
       [-0.18181818,  0.09090909,  0.18181818]])

In [10]:
#Program 12
arr = np.array([1,4,9,16,25])
sqrt_arr = np.sqrt(arr)
sqrt_arr

array([1., 2., 3., 4., 5.])

In [11]:
def custom_sqrt(x):
    return x**0.5

vectorized_sqrt = np.vectorize(custom_sqrt)
sqrt_arr2 = vectorized_sqrt(arr)
sqrt_arr2


array([1., 2., 3., 4., 5.])

In [12]:
matrix = np.sqrt(A)
matrix

array([[1.        , 1.41421356, 1.73205081],
       [0.        , 2.        , 2.23606798],
       [1.        , 0.        , 2.44948974]])

In [13]:
import numpy as np

# Define two 2D arrays (matrices)
A = np.array([[1, 2, 3],
              [4, 5, 6]])

B = np.array([[6, 5, 4],
              [3, 2, 1]])

print("Matrix A:\n", A)
print("Matrix B:\n", B)

# 1️⃣ Addition
C_add = A + B
print("\nA + B =\n", C_add)

# 2️⃣ Subtraction
C_sub = A - B
print("\nA - B =\n", C_sub)

# 3️⃣ Element-wise Multiplication
C_mul = A * B
print("\nA * B (element-wise) =\n", C_mul)

# 4️⃣ Element-wise Division
C_div = A / B
print("\nA / B (element-wise) =\n", C_div)

# 5️⃣ Matrix Multiplication (Dot Product)
C_dot = np.dot(A, B.T)  # B.T = transpose to match dimensions
print("\nA dot B^T =\n", C_dot)


Matrix A:
 [[1 2 3]
 [4 5 6]]
Matrix B:
 [[6 5 4]
 [3 2 1]]

A + B =
 [[7 7 7]
 [7 7 7]]

A - B =
 [[-5 -3 -1]
 [ 1  3  5]]

A * B (element-wise) =
 [[ 6 10 12]
 [12 10  6]]

A / B (element-wise) =
 [[0.16666667 0.4        0.75      ]
 [1.33333333 2.5        6.        ]]

A dot B^T =
 [[28 10]
 [73 28]]


In [14]:
import pandas as pd

# Example dataset
data = {
    'Name': ['Alice', 'Bob', 'Charlie', 'David', None],
    'Age': [25, None, 30, 22, 28],
    'City': ['New York', 'Los Angeles', None, 'Chicago', 'Houston']
}

# Create a DataFrame
df = pd.DataFrame(data)

print("Dataset:\n")
print(df)

# Check for missing values
print("\nMissing values (True = missing):\n")
print(df.isnull())

# Count missing values in each column
print("\nNumber of missing values in each column:\n")
print(df.isnull().sum())

# Total missing values in the dataset
print("\nTotal missing values in the dataset:", df.isnull().sum().sum())


Dataset:

      Name   Age         City
0    Alice  25.0     New York
1      Bob   NaN  Los Angeles
2  Charlie  30.0         None
3    David  22.0      Chicago
4     None  28.0      Houston

Missing values (True = missing):

    Name    Age   City
0  False  False  False
1  False   True  False
2  False  False   True
3  False  False  False
4   True  False  False

Number of missing values in each column:

Name    1
Age     1
City    1
dtype: int64

Total missing values in the dataset: 3


In [25]:
#Problem 15
import pandas as pd

# Example dataset
data = {
    'Department': ['Sales', 'HR', 'Sales', 'IT', 'HR', 'IT', 'Sales'],
    'Employee': ['Alice', 'Bob', 'Charlie', 'David', 'Eva', 'Frank', 'Grace'],
    'Salary': [50000, 60000.2, 55000, 70000.3456, 65000, 72000, 58000],
    'Experience': [2, 5, 3, 7, 6, 8, 4]
}

# Create DataFrame
df = pd.DataFrame(data)
print("Original Dataset:\n")
print(df)

# Group by 'Department' and calculate sum of Salary and average Experience
grouped = df.groupby('Department').agg({
    'Salary': 'sum',
    'Experience': 'mean'
})

print("\nGrouped and Aggregated Data:\n")
print(grouped)

# You can also get count of employees per department
count_group = df.groupby('Department')['Employee'].count()
print("\nNumber of employees in each department:\n")
print(count_group)


Original Dataset:

  Department Employee      Salary  Experience
0      Sales    Alice  50000.0000           2
1         HR      Bob  60000.2000           5
2      Sales  Charlie  55000.0000           3
3         IT    David  70000.3456           7
4         HR      Eva  65000.0000           6
5         IT    Frank  72000.0000           8
6      Sales    Grace  58000.0000           4

Grouped and Aggregated Data:

                 Salary  Experience
Department                         
HR          125000.2000         5.5
IT          142000.3456         7.5
Sales       163000.0000         3.0

Number of employees in each department:

Department
HR       2
IT       2
Sales    3
Name: Employee, dtype: int64


In [26]:
grouped_multi = df.groupby('Department')['Salary'].agg(['sum','mean','max'])
print("\nMultiple aggregations on Salary:\n", grouped_multi)


Multiple aggregations on Salary:
                     sum          mean      max
Department                                    
HR          125000.2000  62500.100000  65000.0
IT          142000.3456  71000.172800  72000.0
Sales       163000.0000  54333.333333  58000.0


In [27]:
grouped_multi['mean'] = grouped_multi['mean'].round(2)
grouped_multi['sum'] = grouped_multi['sum'].round(3)
grouped_multi['max'] = grouped_multi['max'].round(2)
print(grouped_multi)

                   sum      mean      max
Department                               
HR          125000.200  62500.10  65000.0
IT          142000.346  71000.17  72000.0
Sales       163000.000  54333.33  58000.0
